### PR TITLE
Jwks

### PR DESCRIPTION
**Pull request by antonioT90**
### List of changes

jwks file generation with jwt certificate

### Type of changes

- [x] Add new resources
- [ ] Update configuration to existing resources
- [ ] Remove existing resources

### Env to apply

- [x] DEV
- [ ] UAT
- [ ] PROD

### Does this introduce a change to production resources with possible user impact?

- [ ] Yes, users may be impacted applying this change
- [x] No

### Does this introduce an unwanted change on infrastructure? Check terraform plan execution result

- [ ] Yes
- [x] No

### Other information

---

### CODE REVIEW

File: src/core/security.tf [hunk 1]
```diff
@@ -231,3 +231,76 @@ resource "azurerm_key_vault_secret" "jwt_public_key" {
...
+
+resource "null_resource" "upload_jwks" {
+  triggers = {
+    "changes-in-jwt" : azurerm_key_vault_certificate.jwt_certificate.thumbprint
+  }
+  provisioner "local-exec" {
+    command = <<EOT
+              mkdir -p "${path.module}/.terraform/tmp"
+              pip install authlib
+              jwk=$(python "${path.module}/utils/py/jwkFromPem.py" "${format("%s\n%s\n%s", "-----BEGIN CERTIFICATE-----", replace(azurerm_key_vault_certificate.jwt_certificate.certificate_data_base64, "/(.{64})/", "$1\n"), "-----END CERTIFICATE-----")}")
```
Comment (pasqualedevita) [inline src/core/security.tf:273]: is possibile to use openssl commands?

Comment (antonioT90) [reply src/core/security.tf:273]: it's difficutl build the required infos (modulus and exponent of the rsa algorithm encoded as asn1) using openssl. this python script is the simplest method found

File: src/core/security.tf [hunk 2]
```diff
@@ -231,3 +231,76 @@ resource "azurerm_key_vault_secret" "jwt_public_key" {
...
+    issuer_parameters {
+      name = "Self"
+    }
+
+    key_properties {
+      exportable = true
+      key_size   = 2048
+      key_type   = "RSA"
+      reuse_key  = true
+    }
```
[the remark this answers is not in the document]
Comment (pasqualedevita) [reply src/core/security.tf:255]: @pp-ps have u a suggestion for algorithm/key lenght?

[the remark this answers is not in the document]
Comment (pp-ps) [reply src/core/security.tf:255]: As long as the call to forge a token (signed with this key) is authenticated and observable in metrics, I don't have any strong bias against RSA. EC verification (for receiveing systems) is much more expensive.

RSA 2048 bit looks fine too, as long as the key is periodically rotated (1y?, let's think about if as an expiring certificate)

What I find strange is the choice to pack everything in a azurerm_key_vault_certificate. Why can't we use a azurerm_key_vault_key?

File: src/core/security.tf (hunk 1 above)
Comment (pp-ps) [reply src/core/security.tf:273]: This approach doesn't take into consideration a potential rotation and coexistence of more keys for a time window.
During a rotation the JWKS will be a set of JWK (new and old key), the code will use the first key to sign new tokens, while clients still holding older tokens will be able to talk to the server (old key still there).

File: src/core/security.tf (hunk 2 above)
[the remark this answers is not in the document]
Comment (pasqualedevita) [reply src/core/security.tf:255]: @antonioT90 can u try to use azurerm_key_vault_key?
with azurerm_key_vault_key u can retrieve 
n - The RSA modulus of this Key Vault Key.
e - The RSA public exponent of this Key Vault Key.

https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/key_vault_key#attributes-reference

[the remark this answers is not in the document]
Comment (antonioT90) [reply src/core/security.tf:255]: @pasqualedevita i'm not able to retrieve the private key generated using key_vault_key, we need it in order to firm JWT token in hubSpidLogin application

[the remark this answers is not in the document]
Comment (antonioT90) [reply src/core/security.tf:255]: same answer to @pp-ps! the use of key_vault_key will not allow me to read the private key generated

the use of azurerm_key_vault_certificate is due to transform the private and public keys into a X509 certificate, simplifying the calculation of thumbprint

[the remark this answers is not in the document]
Comment (antonioT90) [reply src/core/security.tf:255]: regarding the rotation of the keys, there is a problem due to the inability to replace a previous version of a key: actually I obtain an error due to the safe delete feature enabled which will not allow to perform a delete and purge of previous value...

File: src/core/security.tf [hunk 3]
```diff
@@ -231,3 +231,76 @@ resource "azurerm_key_vault_secret" "jwt_public_key" {
...
+}
+
+resource "null_resource" "upload_jwks" {
+  triggers = {
+    "changes-in-jwt" : azurerm_key_vault_certificate.jwt_certificate.thumbprint
+  }
+  provisioner "local-exec" {
+    command = <<EOT
+              mkdir -p "${path.module}/.terraform/tmp"
+              pip install authlib
```
[the remark this answers is not in the document]
Comment (antonioT90) [reply src/core/security.tf]: resolved in new version (I refactored the code, so all the previous files changed)

File: src/core/security.tf (hunk 1 above)
Comment (antonioT90) [reply src/core/security.tf]: in the new logic, the jwks will take into account the previous file

**Review by pasqualedevita — APPROVED** [review]
LGTM, @pp-ps ?

**Comment by pp-ps** [comment]
I hear we need to speed up deployment of this, let's go ahead and refactor later.

- line 46 in jwksFromPems.py is way too complex
- if we move away from self-signed certs, we can avoid PEMs and so x5c and x5t (use raw keys)

Approving for now.

**Review by pp-ps — APPROVED** [review]

**Pull request merged**